lib/types.ts

```diff
@@ -93,15 +93,26 @@ export type DeepWritable<T> = T extends Primitive
 interface WritableSet<ItemType> extends Set<DeepWritable<ItemType>> {}
 interface WritableMap<KeyType, ValueType> extends Map<DeepWritable<KeyType>, DeepReadonly<ValueType>> {}
 
-/**
- * Omit given key in object type
- * @deprecated Starting with TypeScript 3.5, Omit is natively available.
- */
-export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
-
 /** Omit all properties of given type in object type */
 export type OmitProperties<T, P> = Pick<T, { [K in keyof T]: T[K] extends P ? never : K }[keyof T]>;
 
+/** Recursively omit deep properties */
+export type DeepOmit<T, P extends {}> = T extends Primitive | Function | Date
+  ? T
+  : T extends Map<infer K, infer V>
+  ? DeepOmitMap<K, V, P>
+  : T extends Set<infer U>
+  ? DeepOmitSet<U, P>
+  : T extends Array<infer U>
+  ? DeepOmitArray<U, P>
+  : T extends {}
+  ? { [K in Exclude<keyof T, keyof P>]: T[K] } &
+      OmitProperties<{ [K in Extract<keyof T, keyof P>]: P[K] extends true ? never : DeepOmit<T[K], P[K]> }, never>
+  : T;
+interface DeepOmitArray<ItemType, P> extends Array<DeepOmit<ItemType, P>> {}
+interface DeepOmitSet<ItemType, P> extends Set<DeepOmit<ItemType, P>> {}
+interface DeepOmitMap<KeyType, ValueType, P> extends Map<DeepOmit<KeyType, P>, DeepOmit<ValueType, P>> {}
```

Comment:
I don't like the fact that the same type filter is applied to both keys and values.
Maybe filter only values? Indexing maps with objects is probably a rare case.
Maybe write the type filter for `Map` separately for keys and values?

Or a combination of the above: using just one type in the filter filters values, if you want to filter keys too, use `Map<KeyFilter, ValueFilter>`:

```typescript
type DeepOmitMap<KeyType, ValType, Filter> =
  Filter extends Map<infer KeyFilter, infer ValFilter>
    ? Map<DeepOmit<KeyType, KeyFilter>, DeepOmit<ValType, ValFilter>>
    : Map<KeyType, DeepOmit<ValType, Filter>>
```

This looks nice, but maybe I'm going too far for the sake of type purity? :joy_cat: Do you have any ideas to avoid using the same type filter for keys & values?

Reply:
CC @macbem

Reply:
For now, I prefer to only apply the filter to the value of `Map`.